theseus/core/vectorizer.py

```diff
@@ -0,0 +1,291 @@
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple
+
+import torch
+
+from .cost_function import CostFunction, _register_vars_in_list
+from .objective import Objective
+from .variable import Variable
+
+_CostFunctionSchema = Tuple[str, ...]
+
+
+def _get_cost_function_schema(cost_function: CostFunction) -> _CostFunctionSchema:
+    def _fullname(obj) -> str:
+        return f"{obj.__module__}.{obj.__class__.__name__}"
+
+    def _varinfo(var) -> str:
+        return f"{_fullname(var)}{tuple(var.shape[1:])}"
+
+    return (
+        (_fullname(cost_function),)
+        + tuple(_varinfo(v) for v in cost_function.optim_vars)
+        + tuple(_varinfo(v) for v in cost_function.aux_vars)
+        + (_fullname(cost_function.weight),)
```

Comment:
Thought no strong preferences, I think it is better to keep it possible to vectorize cost functions of different types of weights. In most cases, `CostFunction.errors()` and `CostFunction.jacobians()` are the most time-consuming.

Reply:
I agree this is desirable to have, for the reason you mention. That being said, it seems OK to me to open an issue for this and revisit later, once we have more feedback from users, because of two reasons:

1. separating into error/jacobian and weight makes the code more complicated, 
2. more important, it seems to me that it will be a rare use case to have a bunch of cost functions of the same type but with different cost weight types, so we might be over engineering for an atypical use case.

Reply:
Makes sense. Let's open an issue for this to tackle later.

Reply:
Created #208